File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.cc
#include <bits/stdc++.h>

#define MX_W 100'002
#define MX_N 101

using namespace std;

int mem[MX_N][MX_W];
pair<int, int> items[MX_N];
int n, k;

int dp(int idx, int w) {
    if (idx == 0) return 0;

    if (mem[idx][w] != -1) return mem[idx][w];

    mem[idx][w] = dp(idx - 1, w);

    if (w >= items[idx].first) {
        mem[idx][w] = max(
                mem[idx][w],
                items[idx].second + dp(idx - 1, w - items[idx].first)
        );
    }

    return mem[idx][w];
}

void init() {
    cin >> n >> k;
    for (int i = 1; i <= n; i++) {
        int w, v;
        cin >> w >> v;
        items[i] = {w, v};
    }

    memset(mem, -1, sizeof(mem));
}

void solve() {
    cout << dp(n, k) << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    init();
    solve();
}

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.*;
import java.util.*;

public class Main {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    public static int N;
    public static int K;

    public static int[][] items;
    public static int[][] mem;

    public static void init() throws Exception {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        items = new int[N + 1][2];
        for (int i = 1; i <= N; i++) {
            st = new StringTokenizer(br.readLine());
            items[i][0] = Integer.parseInt(st.nextToken());
            items[i][1] = Integer.parseInt(st.nextToken());
        }
        mem = new int[N + 1][K + 1];
    }

    public static void solve() {
        for (int i = 1; i <= N; i++) {
            int w = items[i][0];
            int v = items[i][1];
            for (int j = 1; j <= K; j++) {
                if (w > j) {
                    mem[i][j] = mem[i - 1][j];
                    continue;
                }
                mem[i][j] = Math.max(v + mem[i - 1][j - w], mem[i - 1][j]);
            }
        }

        System.out.print(mem[N][K]);
    }


    public static void main(String[] args) throws Exception {
        init();
        solve();
    }
}
